### ISSUE TITLE
Context cancel not working on blocking operations in v9

### ISSUE THREAD

**Issue by FLAGLORD** [redis/go-redis#2556]
I noticed the https://github.com/redis/go-redis/issues/1196#issue-527462854, and it seems that maintainers have fixed this.

But i use the alomost the same code given in the https://github.com/redis/go-redis/issues/1196#issue-527462854:
```go
client := redis.NewClient(&redis.Options{
		Addr: "localhost:6379",
	})

cancelCtx, cancelFn := context.WithCancel(context.Background())

wg := &sync.WaitGroup{}
wg.Add(1)

go func() {
	client.BRPop(cancelCtx, 0, "")
	wg.Done()
}()

time.Sleep(2 * time.Second)

cancelFn()
wg.Wait()
```
It still waits indefinitely due to the 0 timeout. I have no idea...

**Comment by FLAGLORD** [comment]
Is it related to https://github.com/redis/go-redis/issues/2276 ?

**Comment by SoulPancake** [comment]
@FLAGLORD I think this might be related to
https://redis.io/commands/blpop/#:~:text=A%20timeout%20of%20zero%20can%20be%20used%20to%20block%20indefinitely.
Most probably the same applies to BRPOP even though I don't see it in the updated documentation.

**Comment by monkey92t** [comment]
We made an attempt in #2455, to be honest, the experience was poor. It disrupted our usual process of executing commands. In my opinion, we need to put more thought into it.

**Comment by monkey92t** [comment]
I feel that the cost of #2455 is too high, and I don't think it's a good solution, even though it solves the problem.

**Comment by agorman** [comment]
Is there a reasonable solution to this problem other than setting a low timeout and checking the value of context manually? Seems like we don't have a choice but to poll which defeats the purpose of the blocking pop. It also means I have to wait up to a second for it to respond to a context being canceled.

Anyone have a nicer solution?

~~~
var value string
for {
	select {
	case <-ctx.Done():
		return nil, context.Canceled
	default:
	}

	result, err := client.BLPop(ctx, time.Second, key).Result()
	if err == nil {
		value = result[1]
		break
	}
}
~~~

**Comment by FLAGLORD** [comment]
> 

If you hope that it returns immediately when being canceled, code snippet below works.
```go
var value string
valChan := make(chan string)
// Use closeCtx to avoid goroutine leak
closeCtx, closeCtxCancel := context.WithCancel(ctx)

go func() {
	for {
		select {
		case <-closeCtx.Done():
			return
		default:
		}
		result, err := client.BLPop(ctx, time.Second, key).Result()
		if err == nil {
			valChan <- result[1]
			return
		}
	}
}()

{
	defer closeCtxCancel()
	select {
	case <-ctx.Done():
		return nil, context.Canceled
	case value = <-valChan:
	}
}
```
However, it seems weird and tricky. **I do not like it.**

**Comment by FLAGLORD** [comment]
Sadly my code snippet has side effects: 
If the program successfully read result from redis, but it hasn't notified by writing to `donnChan`. At this time, `ctx` is canceled. It will return context.canceled but has retrieved value from redis...

**Comment by FLAGLORD** [comment]
It seems impossible that there could be a solution that has no side effects and responds to context cancel quickly. Considering that waiting for cancel and `BLPop` are independent from each other:

-  If the program responds to cancel immediately, we have to use a goroutine to do `BLPop`, and a cancel can be sent while blocking(it means side effects);
- If we hope to avoid dropping `BLPop` results, we have to block in the main process and it means that we have to tolerate a delay in context cancel.

Currently we have to trade off according to our needs

**Comment by monkey92t** [comment]
Because conn.Read(), conn.Write() do not depend on ctx, but only use timeout to control.

**Comment by abemedia** [comment]
> Because conn.Read(), conn.Write() do not depend on ctx, but only use timeout to control.

You could simply using `context.AfterFunc` to set the deadline:

```go
done := make(chan struct{})
stop := context.AfterFunc(ctx, func() {
	conn.SetReadDeadline(time.Now())
	close(stopc)
})
n, err = conn.Read(b)
if !stop() {
	<-done // Wait for AfterFunc to complete.
	conn.SetReadDeadline(time.Time{})
	return n, ctx.Err()
}
```

This might not work for calls with side-effects but would work nicely for things like `XRead` and it's a lot more lightweight than the solution in the PR above.

**Comment by abemedia** [comment]
> Because conn.Read(), conn.Write() do not depend on ctx, but only use timeout to control.

You could simply using `context.AfterFunc` to set the deadline:

```go
done := make(chan struct{})
stop := context.AfterFunc(ctx, func() {
	conn.SetReadDeadline(time.Now())
	close(done)
})
n, err = conn.Read(b)
if !stop() {
	<-done // Wait for AfterFunc to complete.
	conn.SetReadDeadline(time.Time{})
	// Return `conn` to pool.
	return n, ctx.Err()
}
```

This might not work for calls with side-effects but would work nicely for things like `XRead` and it's a lot more lightweight than the solution in the PR above.

It could also be further optimised by using a `sync.Pool` for for the channels that are created for waiting.

**Comment by schnz** [comment]
I just ran into the same issue with a blocking `XREADGROUP` command. After reading through this issue, I wasn't quite satisfied with the workarounds offered here. After further investigation and a little digging into the source of this library, I resorted to a solution that wasn't yet suggested here or elsewhere. I'll outline the idea below:

I resorted to use a dedicated connection for blocking commands and use the connection ID to unblock the client whenever my context closes.

Code example:
```go
var rdb *redis.Client

ctx, cancel := context.WithTimeout(context.Background(), 15 * time.Second)
defer cancel()

conn := rdb.Conn()
connectionID := conn.ClientID(ctx).Val()

stop := ctx.AfterFunc(ctx, func()  {
    // uses one of the pooled connections of the redis client a to unblock the  blocking connection
    rdb.ClientUnblock(context.Background(), connectionID)
})
defer stop()

// imagine this to be run in a for{} loop which is left once ctx.Done() is closed.
// to check if the channel was closed, XReadGroup must return prematurely if that happens
result, err := conn.XReadGroup(ctx, &redis.XReadGroupArgs{
    Group:    "mygroup",
    Consumer: "myclient",
    Streams:  []string{"mystream", ">"},
    Block:    10 * time.Second,
}).Result()

// returns immediately when the group has new messages
// also returns immediately when ctx.Done() is closed (because of the issued `CLIENT UNBLOCK` command)
// blocks otherwise until timeout (10s) is reached or new messages arrive
```

**Comment by jimsnab** [comment]
In my case, all I wanted from a canceled context was graceful termination. So I worked around this problem with the following:

```
go func() {
  <-ctx.Done()
  rdcli.Close()
}()
```

This rudely closes the underlying net Conn and causes something like `BLMove` to return err set to `context.Canceled`.

It remains possible that the blocking operation had just completed its operation on the redis server and the result was on its way to be processed but the socket was shut and the response was lost. Since I've designed protections for abnormal process termination already, this was not an issue for me.

**Comment by arkadijs** [comment]
For Redis cluster client this issue is even worse as can't use `CLIENT UNBLOCK` via current API surface.

**Comment by ndyakov** [comment]
Hello @arkadijs,
Can you share more details? You should be able to use `CLIENT UNBLOCK` with `ClusterClient.Do` https://github.com/redis/go-redis/blob/d0f921357d9751a3e11221a7abd072546ac67654/osscluster.go#L955

**Comment by arkadijs** [comment]
Yes but to unblock who? @ndyakov 

In my use case it's concurrent BZPOPMIN-s on a specific key and a simple workaround is in place to re-queue _ordered events_ dequeued by the connection of canceled context after the command returns (no reader anymore).

The unblock can be done but it's also race condition -- clustered client or not.

**Comment by monkey92t** [comment]
Please allow me to introduce the process of executing commands in go-redis: cmd -> get conn (mutex) -> conn pool -> net.conn -> check conn -> write cmd -> read resp -> put conn (mutex). Throughout this process, the write/read commands are the most time-consuming operations (ignoring the scenario of establishing a new network connection here). However, during the write/read process, Go (or the operating system API) does not support passing cancellation via context (context includes timeout cancellation and manual cancellation); it only supports setting a timeout duration (write/read timeout). If the context is forcibly canceled during the write/read process, it indicates that the connection is contaminated (having been written with a command and awaiting a response to be read). The current approach is to close the contaminated connection. If there are a large number of context timeouts, this could cause the connection pool to continuously create new connections, potentially leading to a large-scale failure.
﻿
We have tried introducing context judgment, but the cost was prohibitively high. Although it solved the problem, it wasn’t a good solution—akin to building an elevator from Earth to the Moon. While it might address the issue of human lunar travel, it’s fundamentally not a practical approach.
﻿
Additionally, I’ve considered various alternatives. For instance, instead of simply closing contaminated connections, we could manage them separately, purify them, and once they’re clean, allow them to rejoin the connection pool to handle tasks. However, this would exceed the PoolSize limit, and if users are unaware of what’s happening, it could lead to unnecessary misunderstandings. Moreover, this solution hasn’t been implemented or tested—it’s merely a concept. I feel it’s not very reliable and could potentially cause widespread connection pool failures.
﻿
I once envisioned a more complex solution where each connection is managed using a Go channel (go-chan). Commands and results would be passed through channels, and connections would operate like buses in conjunction with the connection pool, using a pipelined send/receive command approach. Simply put, each command corresponds to a channel, while the connection acts like a bus. When a command needs to be executed, it “boards the bus” (the connection). When the connection responds, it reads the result set back to each channel. If a network issue occurs, all channels corresponding to that connection fail.
﻿
Since each command has its own go-chan, handling context (ctx) becomes much more manageable. The connection would no longer need to be forcibly closed due to the cancellation of one or more command contexts, as its relationship with each “passenger” (command) is no longer tightly coupled.
﻿
This solution is overly complex. I thought about it two or three years ago and planned to conduct some tests, but due to time constraints, I never got around to it. Additionally, the changes required would be too extensive and could potentially undermine the current stability of go-redis.

In addition to this, regarding the blocking commands raised in this ISSUE, we need to think more carefully about whether adopting the bus approach might introduce even more problems.

**Comment by vtermanis** [comment]
> This solution is overly complex

I agree that this might not be worth it to build in. But: Why not give your users the tools to do it themselves, as I proposed at the bottom [of this comment](https://github.com/redis/go-redis/issues/2556#issuecomment-1896224021)?
Then they have the choice to deal with the complexity themselves if they deem it important but the core code doesn't have to care and just provides some hooks.

**Comment by arkadijs** [comment]
I hear you @monkey92t, please consider using `client unblock` in the lib to implement context cancelation.

Suggesting a big fat working in documentation on Context limitations in go-redis until then. For me, I spent few hours to rework a service to go-redis to just find out the contexts doesn't work. Cancelations was the reason for the rework in the first place.

Personally, I have very good experience partitioning Go apps into actor-like execution blocks where channels allows for good reasoning, proper timeout and reliable control / shutdown. The network comms would overshadow any overhead channels might introduce.

**Comment by monkey92t** [comment]
I’ve reviewed the content above, and the solution provided in https://github.com/redis/go-redis/issues/2556#issuecomment-1694527802 isn’t particularly ideal, even though it appears to address the issue. As I mentioned, I’ve tried a similar approach to monitor the context, but we still can’t guarantee the health of the connection (conn). This approach disrupts a healthy connection pool—we’ve encountered numerous issues of this kind in the past, with many user reports indicating that canceling the context (ctx) leads to a surge of new connections.
﻿
Currently, our need isn’t simply to add a context timeout. In go-redis v8, this was already implemented (monitoring the context), and attempts were made to use the context’s Deadline as the WriteDeadline or ReadDeadline for network connections, but it resulted in disastrous outcomes. As I pointed out earlier, after a context timeout, we need to additionally handle the health of the connection. The connection might have already been written to or contain residual unread data, which would affect the next connection being reused.
﻿
For this issue, I believe there are many aspects we need to consider. We still face the following unresolved challenges:
﻿
How do we ensure the health of the connection after a context timeout or manual cancellation? If we forcefully close it, we risk entering a vicious cycle of creating a large number of new connections.
How can we implement this at a low cost? In the code from https://github.com/redis/go-redis/issues/2556#issuecomment-1694527802, a goroutine is created for each command to monitor it. This means we incur the overhead of goroutine creation, destruction, and GMP scheduling for every command. In high-load scenarios, this could introduce bottlenecks or reduce system throughput, even though net/http adopts a similar approach.
How do we handle timeouts during blocking commands? Should we rely on the context, the timeout directive, or the Option.Timeout setting?
Stability and reliability: How can we ensure that large-scale context timeouts don’t disrupt normal operations?
Respecting existing configurations: Settings like PoolSize, WriteTimeout, and ReadTimeout should be honored. If significant changes are needed, should we consider introducing them in go-redis v10?
Considering the issues above, it’s not that we’re avoiding solving this problem—it’s about designing a reliable solution that addresses these known challenges. In my view, the cmd->chan->conn bus approach remains a viable option, though it’s complex and involves substantial changes.
﻿
If anyone has better suggestions or would like to contribute a PR, they’re warmly welcomed!

**Comment by monkey92t** [comment]
> Personally, I have very good experience partitioning Go apps into actor-like execution blocks where channels allows for good reasoning, proper timeout and reliable control / shutdown. The network comms would overshadow any overhead channels might introduce.

I agree with your perspective. The context plays a very important role in Go, and as a Go package, we should adhere to it. However, there’s a conflict between the context and network read/write operations, and we should find a peaceful solution to resolve it.

**Comment by arkadijs** [comment]
In a robust app, there will be some form of channel-based frontend before go-redis (or any other comm lib) so that the choice of control is in my own hands and appropriate for the use case -- regardless of lib internals (that may also change).
I think, the scope of this particular pain point is exactly and only blocking Redis operations. Solving all kind of network issues is too much to ask of the lib. If network is broken then the app is likely not functional or severely degraded anyway. It must reply properly with Bad Gateway or similar protocol-specific, cleanup, and that's it.

Please open users to do `client unblock` or `SetReadDeadline()` on the connection reliably (with a choice to close it) -- or something else, and it will be good enough. Some practical solution is better than ideal one never materialized.

**Comment by marianososto** [comment]
As a workaround:
Create a separate stream which you'll use to signal context cancellation, lets call it `EOS-Stream` (**E**nd**O**f**S**tream). Then, in the `XReadGroup` you have to specify your stream and _EOS-Stream_.  Whenever `ctx.Done()` , you delete `EOS-Stream`, that way the XReadGroup unblocks.


I put it also in this issue (https://github.com/redis/go-redis/issues/2276#issuecomment-2872956466)

**Comment by ndyakov** [comment]
Bringing this up to be decided and worked on soon.

**Comment by MrEcco** [comment]
+billion, canceling blocking read would come in handy. I'm currently experiencing issues with it, can confirm that cancellation doesn't work, at least for `BRPOP`. My use case (correct me if I do something nasty):
* Have multiple pods which keep numerous websocket connections
* Pods use BRPOP call to get messages for websockets (timeout is 15 seconds, which is too small for my case, but can handle it yet)
* Another backend app queries one (indefinite) pod and expects to use the specific websocket connection
* On this query, the pod uses LPUSH to specific key, so normally it comes right where it should. It's a kind of distributed go-style channels, via Redis
* In case of websocket reconnection, the new BRPOP starts, but **the previous BRPOP can steal the message**.

So, the basic Redis broker feature doesn't work here. I don't like the idea of writing the code which gets the message back to Redis in case of "stolen" message, but it doesn't leave me a chance

**Comment by arkadijs** [comment]
There is always a race condition - cancellation working properly or not. The context might be canceled but message already dequeued. We put them back in that case.

**Comment by WGH-** [comment]
Context cancelation will always be prone to such bugs. This applies to basically every network client library. I saw pgx (PostgreSQL client library) `Commit(ctx)` being canceled so it returned "context canceled" error, but transaction still made it. If you cancel HTTP request, you never know whether your request hit the server. And so on and so on.

I think just by principle of least surprise go-redis should just abort the connection on the spot, just like everyone else does. Something fancier is responsibility of the user.

Helpers to call `client unblock` behind the scenes would be cool, but I don't think that should be the default, especially given cluster limitations. As a side note, [pgx provides some](https://pkg.go.dev/github.com/jackc/pgx/v5@v5.7.6/pgconn#hdr-Context_Support).

**Comment by MrEcco** [comment]
Totally agree with https://github.com/redis/go-redis/issues/2556#issuecomment-3679642887 and https://github.com/redis/go-redis/issues/2556#issuecomment-3679216543, but the problem is a little bit different :)

```go
func f() {
    conn := redis.NewClient(...)

    ctx, cancel := context.WithCancel(context.Background())
    output := make(chan any)

    // This list doesn't exist, so empty
    listName := "mylist"
    // And during the operation nothing has put into it

    go func() {
        cmd := conn.BRPop(ctx, 15*time.Second, listName)
        if cmd.Err() != nil {
            output <- cmd.Err()
        } else {
            output <- cmd.Val()
        }
    }()

    time.Sleep(1 * time.Second)
    cancel()

    // It takes another 14 seconds to get the response!
    <-output
}
```

Isn't it surprising that if I cancel the context, it DOES NOT actually cancels? BTW I test it on a standalone Redis, no cluster, even no replication. Just a plain lone primary node running locally.

**Comment by WGH-** [comment]
@MrEcco, why "but", we're arguing the same point :)
